drain3/masking.py

```diff
@@ -1,35 +1,72 @@
 """
-Description : This file implements the persist/restore from Kafka
+Description : This file implements the masking using user-provided patterns
 Author      : Moshik Hershcovitch
 Author_email: moshikh@il.ibm.com
 License     : MIT
 """
-import logging
 import re
-from typing import List
+import abc
+from typing import Container
 
-logger = logging.getLogger(__name__)
 
+class AbstractMaskingInstruction(abc.ABC):
 
-class MaskingInstruction:
-    def __init__(self, regex_pattern: str, mask_with: str):
-        self.regex_pattern = regex_pattern
+    def __init__(self, mask_with: str):
         self.mask_with = mask_with
-        self.regex = re.compile(regex_pattern)
 
+    @abc.abstractmethod
+    def mask(self, content: str, mask_prefix: str, mask_suffix: str) -> str:
+        """
+        Mask content according to this instruction and return the result.
 
-class RegexMasker:
-    def __init__(self, masking_instructions: List[MaskingInstruction], mask_prefix: str, mask_suffix: str):
+        :param content: text to apply masking to
+        :param mask_prefix: the prefix of any masks inserted
+        :param mask_suffix: the suffix of any masks inserted
+        """
+        pass
+
+
+class MaskingInstruction(AbstractMaskingInstruction):
+
+    def __init__(self, pattern: str, mask_with: str):
+        super().__init__(mask_with)
+        self.regex = re.compile(pattern)
+
+    @property
+    def pattern(self):
+        return self.regex.pattern
+
+    def mask(self, content: str, mask_prefix: str, mask_suffix: str) -> str:
+        mask = mask_prefix + self.mask_with + mask_suffix
+        return self.regex.sub(lambda x: mask, content)
+
+
+# Alias for `MaskingInstruction`.
+RegexMaskingInstruction = MaskingInstruction
+
+
+class LogMasker:
+
+    def __init__(self, masking_instructions: Container[AbstractMaskingInstruction], mask_prefix: str, mask_suffix: str):
```

Comment:
Containers are not always iterable: https://stackoverflow.com/questions/56989687/are-containers-always-iterable
Please use Iterable instead.